### ISSUE TITLE
AbstractState in Mathematica

### ISSUE THREAD

**Issue by branch171** [CoolProp/CoolProp#2065]
Hi I have CoolProp 6.4.1 high level interface running with Wolfram Mathematica 12.3. 

The required notebook call is: PropsSI = 
  LibraryFunctionLoad["libCoolProp", 
   "PropsSI", {UTF8String, UTF8String, Real, UTF8String, Real, 
    UTF8String}, Real];

Note the need for libCoolProp and not CoolProp as in the example.nb supplied. I used linix (ubuntu) with gcc compiler 9.3

I would like to also use the low level interface. I am happy to try and develop this and test it but I would like to see if anyone has a staring point. It particular the syntax for returning vectors from a function call.

My interest is the need for more robust flashes TP, PH and PS for mixtures using CoolProp HEOS. I have a  test mixture 95% CO2 and 5%N2 and simple dew point and bubble point flashes fail for this mixture as you approach the mixture critical point. The bubble point fails quite a long way from the critical point at 60 bara with Pc = ~80 bara.

That said does the low level interface give me access to partial molar properties like dAres/dni , dGres/dni or partial molar fugacity coefficient. I need these to be able to solve the flash routes my self in Mathematica.

Thanks

**Comment by ibell** [comment]
Do you know if you can wrap classes in Mathematica reasonably easily?  That is what is needed I think, or you could use the low-level accessors from the high-level interface (the ``AbstractState_XXXX`` methods).

No, the low-level does not give you these.  You might want to check out teqp, it has the things you are asking about, with a modern C++ interface : https://github.com/usnistgov/teqp .  But if you wanted to add the necessary accessors to the low-level interface, I would consider a PR.

**Comment by branch171** [comment]
Thanks Ian

I will try and see if you can wrap classes in Mathematica. Will feedback what I find.

As for the partial molar properties. Seems the AbstractState has chemical potential so this is a start

double AbstractState::chemical_potential(std::size_t i)

I will try adding what I need. I'll also try improve the two phase flash which even for simple binary mixture does not converge well from the high level interface. Hence my need to use the low level interface to have more control of things.

**Comment by ibell** [comment]
Yes, it should be possible to expose what is missing, please implement it for the other backends as well (so we can test it).

More input on fixing flash calcs would be welcome. It is a hard problem.

**Comment by friederikeboehm** [comment]
> Thanks Ian
> 
> I will try and see if you can wrap classes in Mathematica. Will feedback what I find.
> 
> As for the partial molar properties. Seems the AbstractState has chemical potential so this is a start
> 
> double AbstractState::chemical_potential(std::size_t i)
> 
> I will try adding what I need. I'll also try improve the two phase flash which even for simple binary mixture does not converge well from the high level interface. Hence my need to use the low level interface to have more control of things.

Hi I am currently also looking into using AbstractState in Mathematica. Have you made progress in this regard? Would be great to exchange ideas on the topic :)

**Comment by branch171** [comment]
Hi I haven't made much progress but will start soon. Once I get going if I have any success I will post.

**Comment by friederikeboehm** [comment]
Hi, I did make quite some progress the last two weeks, adding new functions and writing a Mathematica package that imports these and finally gives usefull error messages (opposed to none at all in the past). I am unsure how to go forward with my contribution, as this is actually my first time contributing on github. I have read the [Guide on contributing](https://github.com/CoolProp/CoolProp/wiki/Contributing:-git-development-workflow) and understand it in general but I do have some questions:

- Do I need to do anything to add documentation about new functions I added to CoolPropLib.h or do I assume correctly that this is done automatically?
- Do I squash my commits to make everything a bit easier to follow (sometimes I only added one small function in a commit, sometimes its like 500 lines of code for one commit) and keep the history shorter?

If this is the wrong place to ask, please redirect me to where I should ask it and I'll remove the comment. I know about the Google Group but that didn't feel like the right place either.

**Comment by ibell** [comment]
I think one PR with edits to coolproplib.h and another with everything else
would be good.   Usually I squash merge when merging into master. The docs
are auto generated with doxygen.

**Comment by ibell** [comment]
Closing as part of an issue-tracker cleanup — this issue has had no activity in over two years.

*To be transparent: this comment and the close action are being performed by Claude (Anthropic's AI coding assistant), running on behalf of @ibell. Ian reviewed and approved the bulk-close criteria; I am the one executing the action.*

If you would like this issue reconsidered, please feel free to **re-open it** — a fresh comment with current behaviour against the latest CoolProp release is the most useful way to revive the discussion.

— Claude (acting for @ibell)
